File: abstractions/dotnet/src/serialization/ParseNodeProxyFactory.cs
// ------------------------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
// ------------------------------------------------------------------------------

using System;
using System.IO;

namespace Microsoft.Kiota.Abstractions.Serialization
{
    /// <summary>
    /// Proxy factory that allows the composition of before and after callbacks on existing factories.
    /// </summary>
    public abstract class ParseNodeProxyFactory : IParseNodeFactory
    {
        /// <summary>
        /// The valid content type for the <see cref="ParseNodeProxyFactory"/> instance
        /// </summary>
        public string ValidContentType { get { return _concrete.ValidContentType; } }
        private readonly IParseNodeFactory _concrete;
        private readonly Action<IParsable> _onBefore;
        private readonly Action<IParsable> _onAfter;
        /// <summary>
        /// Creates a new proxy factory that wraps the specified concrete factory while composing the before and after callbacks.
        /// </summary>
        /// <param name="concrete">The concrete factory to wrap.</param>
        /// <param name="onBefore">The callback to invoke before the deserialization of any model object.</param>
        /// <param name="onAfter">The callback to invoke after the deserialization of any model object.</param>
        public ParseNodeProxyFactory(IParseNodeFactory concrete, Action<IParsable> onBefore, Action<IParsable> onAfter)
        {
            _concrete = concrete ?? throw new ArgumentNullException(nameof(concrete));
            _onBefore = onBefore;
            _onAfter = onAfter;
        }
        /// <summary>
        /// Create a parse node from the given stream and content type.
        /// </summary>
        /// <param name="content">The stream to read the parse node from.</param>
        /// <param name="contentType">The content type of the parse node.</param>
        /// <returns>A parse node.</returns>
        public IParseNode GetRootParseNode(string contentType, Stream content)
        {
            var node = _concrete.GetRootParseNode(contentType, content);
            var originalBefore = node.OnBeforeAssignFieldValues;
            var originalAfter = node.OnAfterAssignFieldValues;
            node.OnBeforeAssignFieldValues = (x) =>
            {
                _onBefore?.Invoke(x);
                originalBefore?.Invoke(x);
            };
            node.OnAfterAssignFieldValues = (x) =>
            {
                _onAfter?.Invoke(x);
                originalAfter?.Invoke(x);
            };
            return node;
        }
    }
}

File: abstractions/typescript/src/serialization/parseNodeProxyFactory.ts
import { Parsable } from "./parsable";
import { ParseNode } from "./parseNode";
import { ParseNodeFactory } from "./parseNodeFactory";

/** Proxy factory that allows the composition of before and after callbacks on existing factories. */
export abstract class ParseNodeProxyFactory implements ParseNodeFactory {
  public getValidContentType(): string {
    return this._concrete.getValidContentType();
  }
  /**
   * Creates a new proxy factory that wraps the specified concrete factory while composing the before and after callbacks.
   * @param _concrete the concrete factory to wrap
   * @param _onBefore the callback to invoke before the deserialization of any model object.
   * @param _onAfter the callback to invoke after the deserialization of any model object.
   */
  constructor(
    private readonly _concrete: ParseNodeFactory,
    private readonly _onBefore: (value: Parsable) => void,
    private readonly _onAfter: (value: Parsable) => void
  ) {
    if (!_concrete) {
      throw new Error("_concrete cannot be undefined");
    }
  }
  public getRootParseNode(
    contentType: string,
    content: ArrayBuffer
  ): ParseNode {
    const node = this._concrete.getRootParseNode(contentType, content);
    const originalBefore = node.onBeforeAssignFieldValues;
    const originalAfter = node.onAfterAssignFieldValues;
    node.onBeforeAssignFieldValues = (value) => {
      this._onBefore && this._onBefore(value);
      originalBefore && originalBefore(value);
    };
    node.onAfterAssignFieldValues = (value) => {
      this._onAfter && this._onAfter(value);
      originalAfter && originalAfter(value);
    };
    return node;
  }
}
